src/value.rs

```diff
@@ -65,43 +66,41 @@ impl NoteValue {
     }
 }
 
-impl Sub for NoteValue {
-    type Output = Result<ValueSum, OverflowError>;
-
-    fn sub(self, rhs: Self) -> Self::Output {
-        let a: i64 = self.0.try_into().map_err(|_| OverflowError)?;
-        let b: i64 = rhs.0.try_into().map_err(|_| OverflowError)?;
-        Ok(ValueSum(a - b))
-    }
-}
-
 /// A sum of Orchard note values.
-#[derive(Clone, Copy, Debug, Default)]
-pub struct ValueSum(i64);
-
-impl ValueSum {
-    /// Creates a value sum from its raw numeric value.
-    ///
-    /// This only enforces that the value is a signed 63-bit integer. Callers should
-    /// enforce any additional constraints on the value's valid range themselves.
-    pub fn from_raw(value: i64) -> Self {
-        ValueSum(value)
+pub trait ValueSum:
+    Clone
+    + Copy
+    + Debug
+    + PartialEq
+    + Add<Output = Option<Self>>
+    + TryFrom<i64>
+    + Into<i64>
+{
+    /// The additive zero for this type.
+    fn zero() -> Self;
+
+    /// Return the difference between two [`NoteValue`] values, checking
+    /// for overflow.
+    fn sub(lhs: NoteValue, rhs: NoteValue) -> Option<Self> {
```

Comment:
This method name is misleading, as it collides with `<ValueSum as Sub>::sub` (which provides the `ValueSum - ValueSum` operator). Name it something like `from_action(spend: NoteValue, output: NoteValue)` instead.